src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+using ArgCheck
+using OffsetArrays
+
+export tmap, tmap!, tmapreduce, treduce
+
+############################## docs ##############################
+function _make_docstring(fname, args, kw=String[])
+    tf = tname(Symbol(fname))
+    sargs = join(args, ", ")
+    skw = join([kw; "[batch_size::Integer]"], ", ")
+    signature = string(tf, "(", sargs, "; ", skw, ")")
+    """
+        $signature
+
+    Threaded analog of [`Base.$fname`](@ref). See [`Base.$fname`](@ref) for a description of arguments.
+    In order to parallelize the arguments are split into chunks whose size in controlled by the `batch_size` keyword.
+    """
+end
+
+function tname(s::Symbol)
+    Symbol("t", s)
+end
+
+const SYMBOLS_MAPREDUCE_LIKE = [:sum, :prod, :minimum, :maximum]
+for fun in SYMBOLS_MAPREDUCE_LIKE
+    tfun = tname(fun)
+    docstring = _make_docstring(fun, ["[f]", "src::AbstractArray"])
+    @eval begin
+        export $tfun
+        @doc $docstring ->
+        function $tfun end
+    end
+end
+
+@doc _make_docstring(:map!, ["f", "dst::AbstractArray", "srcs::AbstractArray..."]) ->
+function tmap! end
+
+@doc _make_docstring(:map, ["f", "srcs::AbstractArray..."]) ->
+function tmap end
+
+@doc _make_docstring(:mapreduce, ["f", "op", "src::AbstractArray"], ["[init]"]) ->
+function tmapreduce end
+
+@doc _make_docstring(:reduce, ["op", "src::AbstractArray"],  ["[init]"]) ->
+function treduce end
+
+############################## Helper functions ##############################
+struct Batches{V}
+    firstindex::Int
+    lastindex::Int
+    batch_size::Int
+    values::V
+    length::Int
+end
+
+function Batches(values, batch_size::Integer)
+    @argcheck batch_size > 0
+    r = eachindex(values)
+    @assert r isa AbstractUnitRange
+    len = ceil(Int, length(r) / batch_size)
+    Batches(first(r), last(r), batch_size, values, len)
+end
+
+Base.length(o::Batches) = o.length
+Base.eachindex(o::Batches) = Base.OneTo(length(o))
+function Base.getindex(o::Batches, i)
+    @boundscheck (@argcheck i in eachindex(o))
+    start = o.firstindex + (i-1) * o.batch_size
+    stop  = min(start + (o.batch_size) -1, o.lastindex)
+    o.values[start:stop]
+end
+
+function default_batch_size(len)
+    len <= 1 && return 1
+    nthreads=Threads.nthreads()
+    items_per_thread = len/nthreads
+    items_per_batch = items_per_thread/4
+    clamp(1, round(Int, items_per_batch), len)
+end
+
+function Base.iterate(o::Batches, state=1)
+    if state in eachindex(o)
+        o[state], state+1
+    else
+        nothing
+    end
+end
+
+mutable struct _RollingCutOut{A,I<:AbstractUnitRange,T} <: AbstractVector{T}
```

Comment:
If you write `_RollingCutOut{T, A <: AbstractArray{T}, I <: AbstractUnitRange} <: AbstractVector{T}`, there would be no need for the constructor below.

Reply:
How is this better than using a `Base.SubArray`?

Reply:
SubArray is not mutable. While in `_RollingCutOut` I mutate the `eachindex` field, which determines which part of the array is "visible".

Reply:
But why not create a new `view` on the fly instead of rolling out our own subarray-like type?

Reply:
Because view allocates. See for instance [here](https://discourse.julialang.org/t/array-views-becoming-dominant-source-of-memory-allocation/6465) for a discussion.

Reply:
Hmm, I am inclined to live with the allocations and wait for the Julia compiler to get better. But I don't mind this approach either. Just please rename it to `MutableSubArray` or something to make the connection with `Base.SubArray`.

Reply:
Thanks for improving the type parameters of `_RollingCutOut`. How did you come up with that. I was astonished, that julia can discover `T` here.

Reply:
If all the parameters are used in the type annotations of the fields, the default non-parametric constructor in Julia figures out the type parameters from the arguments. Only when a parameter is not used for any of the fields that the default non-parametric constructor fails for the obvious reason and you need to use the parametric one.

Reply:
Sorry for the delay btw, got busy again.

Reply:
Ah thanks! It did not occur to me that julia is clever enough to infer `T` from such an indirect usage in a field.